[Proposal] Multi project multi cluster proposal for GKE deployer

chizhg opened the pull request:
--project and --cluster-name parameters have been modified to accept a list of projects and cluster. This change respects previous functionality with only one project and one cluster. On multi-project cases, cluster names do have to end with "-X", being X a 0-based index to which project they should belong. No other dash character is allowed in the cluster name for parsing purposes and the final cluster name will filter out this indexing postfix.

--projects-requested=int has been included to request N projects from Boskos

Internal data structures keep track of the relations between projects and clusters and the difference instance groups

Network parameter is kept as a single element. On one project cases, it's the network used and on multi-project cases, it is the name of the Shared VPC network that will be created on the first project and to which the rest of the projects will be linked with.

Changed default environment from 'staging' to 'prod'

Succesful regression test performed:
$ go run ~/go/src/github.com/kubernetes-sigs/kubetest2/kubetest2-gke --up --down --num-nodes=1 --project=test-project --cluster-name=testcluster --zone=us-central1-a --network=test-network

amwat commented on kubetest2-gke/deployer/common.go:
```diff
@@ -74,3 +95,22 @@ func (d *deployer) initialize() error {
+func filterMultiProjectClusterName(clusters []string, projectID int) ([]string, error) {
+	result := make([]string, 0)
+	for _, clusterName := range clusters {
+		parts := strings.Split(clusterName, "-")
```
This is a breaking change. The separator should be something that the gke cluster name doesn't allow by default.

chizhg replied:
This function will only be called `if numProjects > 1`, for single project, it will still be using the full name as the cluster name, so it's backward compatible.

amwat replied:
we should allow multi-cluster names to be any possible valid gke cluster name. 
otoh the choice of separator for the project suffix can be anything else.

chizhg replied:
Makes sense.
Changed the separator to `:` which is not allowed in gke cluster names.

amwat commented on kubetest2-gke/deployer/deployer.go:
```diff
@@ -218,61 +227,156 @@ func (d *deployer) Build() error {
+
+	for _, project := range d.projects {
+		// TODO Prepare should be performed per GCP project
+		if err := d.prepareGcpIfNeeded(project); err != nil {
```
we shouldn't naively do this now.
prepare used to 
set up the environment (which should be a one time thing) 
and set the default project (which we shouldn't keep doing before every command) 
instead we should now make sure that all commands specify a `--project` and stop relying on setting a default project now that we have multi project.

chizhg replied:
Changed to only call this function for the first GCP project, which is supposed to be the host project for the multi-project profile.

amwat commented on kubetest2-gke/deployer/deployer.go:
```diff
@@ -354,11 +473,7 @@ func (d *deployer) TestSetup() error {
-		return "", err
-	}
-
+func (d *deployer) Kubeconfig(project, cluster string) (string, error) {
```
This is breaking the deployerwithkubeconfig interface.
The kubeconfig returned here should be a comma separated list of all the clusters created.

chizhg replied:
It looks `kubetest2-gke` does not implement the `DeployerWithKubeconfig` interface, which defines `Kubeconfig() (string, error)` instead of `Kubeconfig(project, cluster string) (string, error)`?

Ref: https://github.com/kubernetes-sigs/kubetest2/blob/master/pkg/types/types.go#L86

amwat replied:
yeah it was previously, the current PR changes that behavior.
Also correction they should be `:` separated not comma separated
https://kubernetes.io/docs/tasks/access-application-cluster/configure-access-multiple-clusters/#linux

chizhg replied:
Fixed the function signature, and used the correct separator for multi kubeconfig files - `:` for linux and MAC, `;` for windows.

amwat commented on kubetest2-gke/deployer/deployer.go:
```diff
@@ -218,61 +227,156 @@ func (d *deployer) Build() error {
+			return err
+		}
+
+		for _, cluster := range d.clustersLayout[project] {
```
multi-cluster creation should definitely be parallelized!

chizhg replied:
Definitely agree.
Changed to use `errgroup` to parallelize the operations.

amwat commented:
also there's another dupe PR https://github.com/kubernetes-sigs/kubetest2/pull/12 can we close either one of these?

chizhg commented:
> also there's another dupe PR #12 can we close either one of these?

Yeah we can close that one.
This PR is originally implemented by Alberto but he didn't sign the CLA and he's OOO this week, so I just copied all his change and created a new one.

amwat commented on kubetest2-gke/deployer/common.go:
```diff
@@ -74,3 +95,22 @@ func (d *deployer) initialize() error {
+	result := make([]string, 0)
+	for _, clusterName := range clusters {
+		parts := strings.Split(clusterName, "-")
+		if len(parts) != 2 {
+			return nil, fmt.Errorf("cluster name does not follow expected format (name-projectIndex): %s", clusterName)
+		}
+		projectIndex, err := strconv.Atoi(parts[1])
+		if err != nil {
+			return nil, fmt.Errorf("cluster name does not follow contain a valid project index (name-projectIndex. E.g: cluster-0): %s", err)
+		}
+		if projectIndex == projectID {
+			// Network field is empty
+			result = append(result, parts[0])
+		}
```
insteading of erroring if the syntax is not provided, a fallback where a random project is chosen from the provided list sounds more flexible?

chizhg replied:
Hmm I'm not sure if that flexibility is desirable, if users decide to use multi-project profile, it should be required to specify which project to use to create which clusters, otherwise the result won't be predictable, e.g. it could be possible to have all clusters randomly created in one project and run into quota issues. So I left this unchanged.

amwat replied:
makes sense let's lave that for now. I was mostly worried about finding out too late that the user supplied syntax is wrong, but as long as we validate it early enough should be fine.

chizhg commented:
> Thanks for adding this functionality!
> 
> /ok-to-test
> 
> High level comments: can you break it down into a couple of commits that add different functionality and also move them to different files (especially for all the network stuff etc), that would make reviewing much easier.
> 
> Also since we don't have automated e2e setup for the gke deployer yet
> can you ensure that the overall flow also works?
> 
> something like `kubetest2 gke --up --down --test=exec -- <some basic testing>`
> 
> First pass comments:

Sorry for making this PR so large... I have moved the new logic to `network.go`, but it would be pretty hard for me to break the change down into multiple commits, as the changes are coupled togther, e.g. if I change the `--project` flag to support multiple, I'll have to change all the places where `d.project` is used.

I have manually tested below commands and they are working fine:
- `kubetest2 gke --up --down --test=exec -- <some basic testing>`
- `kubetest2 gke --up --down --network=test-network --test=exec -- <some basic testing>`
- `kubetest2 gke --up --test=exec -- <some basic testing>`

amwat commented on kubetest2-gke/deployer/deployer.go:
```diff
@@ -182,19 +193,25 @@ var _ types.NewDeployer = New
 
-	flags.StringVar(&d.cluster, "cluster-name", "", "Cluster name. Must be set.")
+	flags.StringSliceVar(&d.clusters, "cluster-name", []string{}, "Cluster names separated by comma. Must be set. "+
+		"For multi-project profile, it should be in the format of clusterA:1,clusterB:1,clusterC:2, where the index means the index of the project.")
```
is it 0-based or 1-based?

chizhg replied:
It's 0-based, I have changed the example to have a project index as 0 to make it less confusing.

amwat commented on kubetest2-gke/deployer/deployer.go:
```diff
@@ -82,20 +85,23 @@ type deployer struct {
 	region            string
-	cluster           string
+	clusters          []string
+	clustersLayout    map[string][]string
```
nit: maybe rename to projectClusterLayout ?
also add a comment explaining the mapping.

chizhg replied:
Done.

amwat commented on kubetest2-gke/deployer/deployer.go:
```diff
@@ -223,56 +240,84 @@ func (d *deployer) Up() error {
 	}
 
+	log.Printf("Environment: %v", os.Environ())
+	eg, _ := errgroup.WithContext(context.Background())
```
nice! but we will probably need to get the Contextwithcancel here to propagate to all the cluster create calls.

chizhg replied:
Done. I have changed to use `context.WithCancel` and call the `cancel` method whenever one cluster creation process fails.

amwat commented on kubetest2-gke/deployer/deployer.go:
```diff
@@ -223,56 +240,84 @@ func (d *deployer) Up() error {
+				args = append(args, subNetworkArgs...)
+				args = append(args, cluster)
+				log.Printf("Gcloud command: gcloud %+v\n", args)
+				if err := runWithOutput(exec.Command("gcloud", args...)); err != nil {
```
we might need to introduce functionality that uses https://golang.org/pkg/os/exec/#CommandContext 
in our internal package and pass the contextwithcancel generated above.

chizhg replied:
Done. Made the change together with the above one.

amwat commented on kubetest2-gke/deployer/network.go:
```diff
@@ -0,0 +1,331 @@
+/*
```
Thanks for moving it to a separate file.
I'm not too aware of the functionality being added here, but mostly looks like it's following: https://cloud.google.com/kubernetes-engine/docs/how-to/cluster-shared-vpc#creating_a_network_and_two_subnets
whigh LGTM

high level nit: let's break down most of these into helper functions (apart from createnetwork and setupnetwork) by removing the deployer receiver and pass whatever is required (projects/network names) as arguments.

chizhg replied:
Yes, it's mostly following that doc.
I have updated most of them into helper functions, as you suggested.

amwat commented:
Yeah that for breaking it into separate files, that definitely helped. 
FYI the individual commits need not be fully contained as long as it makes reviewing easier, i.e. simple adding a couple of methods in each commit is also sufficient.

You'll want to amend the commits though so that they provide relevant context. if not you can also chose to squash commits at the end, upto you.

chizhg commented:
/merge squash

chizhg commented on kubetest2-gke/deployer/common.go:
```diff
@@ -74,3 +96,22 @@ func (d *deployer) initialize() error {
+		}
+		projectIndex, err := strconv.Atoi(parts[1])
+		if err != nil {
+			return nil, fmt.Errorf("cluster name does not follow contain a valid project index (name:projectIndex. E.g: cluster:0): %v", err)
```
Added as the description of the flag:
`For multi-project profile, it should be in the format of clusterA:0,clusterB:1,clusterC:2, where the index means the index of the project.`

chizhg commented on kubetest2-gke/deployer/common.go:
```diff
@@ -41,29 +44,48 @@ func (d *deployer) initialize() error {
+		if numProjects > 1 {
+			for i, project := range d.projects {
+				var err error
+				clusters, err := filterClusterNamesForProject(d.clusters, i)
```
Good point!
I have changed the function to be `buildProjectClustersLayout` which will only iterate the cluster_names once.

chizhg commented on kubetest2-gke/deployer/network.go:
```diff
@@ -0,0 +1,331 @@
+		"--project="+d.projects[0],
+		"--format=value(name)")) != nil {
+		// Assume error implies non-existent.
+		log.Printf("Couldn't describe network %q, assuming it doesn't exist and creating it", d.network)
```
Done. Changed all the `log.Printf` to `klog.V(1).Infof`.

chizhg commented on kubetest2-gke/deployer/network.go:
```diff
@@ -0,0 +1,331 @@
+	if runWithNoOutput(exec.Command("gcloud", "compute", "networks", "describe", d.network,
+		"--project="+d.projects[0],
+		"--format=value(name)")) != nil {
+		// Assume error implies non-existent.
```
This is the original code, it looks we can check that from the error message, but I would prefer not touching it in this PR.
I have added a TODO and will make the change as followups.

chizhg commented on kubetest2-gke/deployer/network.go:
```diff
@@ -0,0 +1,331 @@
+	for i, nr := range d.subnetworkRanges {
+		serviceProject := d.projects[i+1]
+		parts := strings.Split(nr, " ")
+		subnetName := d.network + "-" + serviceProject
```
Done. Added the comment `The subnetwork name is in the format of `[main_network]-[service_project_id]`.`

chizhg commented on kubetest2-gke/deployer/network.go:
```diff
@@ -0,0 +1,331 @@
+	// Create network if it doesn't exist.
+	// Verify networks for all projects and configure the mesh if parameter is set
+	if runWithNoOutput(exec.Command("gcloud", "compute", "networks", "describe", d.network,
+		"--project="+d.projects[0],
```
Done, added the comment for both single-project and multi-project.

chizhg commented:
/retest

chizhg commented:
/retest

chizhg commented:
> Thanks for addressing all the comments!
> Super excited to test multi-project and multi-cluster out.
> 
> /lgtm
> /approve

Thanks a lot for reviewing! 😄

chizhg commented:
/retest
